// @zappar/zappar-cv/lib/native.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { image_target_type_t } from "./gen/zappar";
import { zappar_client } from "./gen/zappar-client";
import { drawPlane } from "./drawplane";
import { cameraRotationForScreenOrientation, projectionMatrix } from "./cameramodel";
import { mat4, vec3 } from "gl-matrix";
import { launchWorker, messageManager } from "./worker-client";
import { permissionRequestUI, permissionGrantedAll, permissionGrantedCamera, permissionGrantedMotion, permissionDeniedAny, permissionDeniedCamera, permissionDeniedMotion, permissionRequestMotion, permissionRequestCamera, permissionRequestAll, permissionDeniedUI } from "./permission";
import { createFaceMesh, destroyFaceMesh, getFaceMesh } from "./facemesh";
import { Pipeline, applyScreenCounterRotation } from "./pipeline";
import { CameraSource } from "./camera-source";
import { HTMLElementSource } from "./html-element-source";
import { createFaceLandmark, destroyFaceLandmark, getFaceLandmark } from "./facelandmark";
import compatibility from './compatibility';
import { setLogLevel, zcwarn, zcout } from "./loglevel";
import { SequenceSource } from "./sequencesource";
import { createCameraSource, getCameraSource } from "./camera-source-map";
import { gfx } from "./gfx";
import { ImageTracker } from "./imagetracker";
let client;
export function initialize(opts) {
    if (client)
        return client;
    let loaded = false;
    launchWorker(opts === null || opts === void 0 ? void 0 : opts.worker).then(() => {
        zcout("Fully loaded");
        loaded = true;
    });
    let c = new zappar_client(ab => {
        messageManager.postOutgoingMessage({
            t: "zappar",
            d: ab
        }, [ab]);
    });
    if (window.location.hostname.toLowerCase().indexOf(".zappar.io") > 0 || window.location.hostname.toLowerCase().indexOf(".webar.run") > 0) {
        let pathParts = window.location.pathname.split("/");
        if (pathParts.length > 1 && pathParts[1].length > 0)
            c.impl.analytics_project_id_set(".wiz" + pathParts[1]);
    }
    messageManager.onIncomingMessage.bind(msg => {
        var _a, _b, _c, _d, _e, _f, _g;
        switch (msg.t) {
            case "zappar":
                (_a = Pipeline.get(msg.p)) === null || _a === void 0 ? void 0 : _a.pendingMessages.push(msg.d);
                break;
            case "buf":
                c.serializer.bufferReturn(msg.d);
                break;
            case "cameraFrameRecycleS2C": {
                let msgt = msg;
                (_c = (_b = Pipeline.get(msgt.p)) === null || _b === void 0 ? void 0 : _b.cameraTokenReturn) === null || _c === void 0 ? void 0 : _c.call(_b, msg);
                break;
            }
            case "videoFrameS2C": {
                let msgt = msg;
                (_e = (_d = Pipeline.get(msgt.p)) === null || _d === void 0 ? void 0 : _d.videoFrameFromWorker) === null || _e === void 0 ? void 0 : _e.call(_d, msgt);
                break;
            }
            case "imageBitmapS2C": {
                let msgt = msg;
                (_g = (_f = Pipeline.get(msgt.p)) === null || _f === void 0 ? void 0 : _f.imageBitmapFromWorker) === null || _g === void 0 ? void 0 : _g.call(_f, msgt);
                break;
            }
            case "licerr": {
                let div = document.createElement("div");
                div.innerHTML = "Visit <a href='https://docs.zap.works/universal-ar/licensing/' style='color: white;'>our licensing page</a> to find out about hosting on your own domain.";
                div.style.position = "absolute";
                div.style.bottom = "20px";
                div.style.width = "80%";
                div.style.backgroundColor = "black";
                div.style.color = "white";
                div.style.borderRadius = "10px";
                div.style.padding = "10px";
                div.style.fontFamily = "sans-serif";
                div.style.textAlign = "center";
                div.style.left = "10%";
                div.style.zIndex = Number.MAX_SAFE_INTEGER.toString();
                let span = document.createElement("span");
                span.innerText = " (30)";
                div.append(span);
                let indx = 30;
                setInterval(function () {
                    indx--;
                    if (indx >= 0)
                        span.innerText = " (" + indx.toString() + ")";
                }, 1000);
                document.body.append(div);
            }
            case "gfx": {
                let div = document.createElement("div");
                div.innerHTML = gfx;
                div.style.position = "absolute";
                div.style.bottom = "20px";
                div.style.width = "250px";
                div.style.left = "50%";
                div.style.marginLeft = "-125px";
                div.style.zIndex = Number.MAX_SAFE_INTEGER.toString();
                div.style.opacity = "0";
                div.style.transition = "opacity 0.5s";
                document.body.append(div);
                setTimeout(function () {
                    div.style.opacity = "1";
                }, 500);
                setTimeout(function () {
                    div.style.opacity = "0";
                }, 3000);
                setTimeout(function () {
                    div.remove();
                }, 4000);
            }
        }
    });
    client = Object.assign(Object.assign({}, c.impl), { loaded: () => loaded, camera_default_device_id: userFacing => userFacing ? CameraSource.USER_DEFAULT_DEVICE_ID : CameraSource.DEFAULT_DEVICE_ID, camera_source_create: (p, deviceId) => createCameraSource(p, deviceId), camera_source_destroy: cam => { var _a; return (_a = getCameraSource(cam)) === null || _a === void 0 ? void 0 : _a.destroy(); }, camera_source_pause: cam => { var _a; return (_a = getCameraSource(cam)) === null || _a === void 0 ? void 0 : _a.pause(); }, camera_source_start: cam => { var _a; return (_a = getCameraSource(cam)) === null || _a === void 0 ? void 0 : _a.start(); }, camera_count: () => 2, camera_id: indx => indx === 0 ? CameraSource.DEFAULT_DEVICE_ID : CameraSource.USER_DEFAULT_DEVICE_ID, camera_name: indx => indx === 0 ? "Rear-facing Camera" : "User-facing Camera", camera_user_facing: indx => indx !== 0, pipeline_create: () => Pipeline.create(c.impl, messageManager), pipeline_frame_update: (p) => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.frameUpdate(c); }, pipeline_camera_frame_draw_gl: (pipeline, screenWidth, screenHeight, mirror) => {
            var _a;
            (_a = Pipeline.get(pipeline)) === null || _a === void 0 ? void 0 : _a.cameraFrameDrawGL(screenWidth, screenHeight, mirror);
        }, draw_plane: (gl, projectionMatrix, cameraMatrix, targetMatrix, texture) => {
            drawPlane(gl, projectionMatrix, cameraMatrix, targetMatrix, texture);
        }, pipeline_draw_face: (p, projectionMatrix, cameraMatrix, targetMatrix, o) => {
            var _a;
            let obj = getFaceMesh(o);
            if (!obj) {
                zcwarn("attempting to call draw_face on a destroyed zappar_face_mesh_t");
                return new Uint16Array();
            }
            (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.drawFace(projectionMatrix, cameraMatrix, targetMatrix, obj);
        }, pipeline_draw_face_project: (p, matrix, vertices, uvMatrix, uvs, indices, texture) => {
            var _a;
            (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.drawFaceProject(matrix, vertices, uvMatrix, uvs, indices, texture);
        }, pipeline_draw_image_target_preview: (p, projectionMatrix, cameraMatrix, targetMatrix, o, indx) => {
            var _a;
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("image tracker not found");
                return;
            }
            (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.drawImageTargetPreview(projectionMatrix, cameraMatrix, targetMatrix, indx, obj);
        }, projection_matrix_from_camera_model: projectionMatrix, projection_matrix_from_camera_model_ext: projectionMatrix, pipeline_process_gl: p => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.processGL(); }, pipeline_gl_context_set: (p, gl, texturePool) => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.glContextSet(gl, texturePool); }, pipeline_gl_context_lost: (p) => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.glContextLost(); }, pipeline_camera_frame_upload_gl: p => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.uploadGL(); }, pipeline_camera_frame_texture_gl: p => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.cameraFrameTexture(); }, pipeline_camera_frame_texture_matrix: (p, sw, sh, mirror) => { var _a; return ((_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.cameraFrameTextureMatrix(sw, sh, mirror)) || mat4.create(); }, pipeline_camera_frame_user_facing: p => { var _a; return ((_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.cameraFrameUserFacing()) || false; }, pipeline_camera_pose_default: () => mat4.create(), pipeline_camera_pose_with_attitude: (p, mirror) => { var _a; return ((_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.cameraPoseWithAttitude(mirror)) || mat4.create(); }, pipeline_camera_pose_with_origin: (p, o) => { let res = mat4.create(); mat4.invert(res, o); return res; }, pipeline_sequence_record_clear: p => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordClear(); }, pipeline_sequence_record_start: (p, expectedFrames) => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordStart(expectedFrames); }, pipeline_sequence_record_stop: p => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordStop(); }, pipeline_sequence_record_device_attitude_matrices_set: (p, v) => { var _a; return (_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordDeviceAttitudeMatrices(v); }, pipeline_sequence_record_data: p => { var _a; return ((_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordData()) || new Uint8Array(0); }, pipeline_sequence_record_data_size: p => { var _a; return ((_a = Pipeline.get(p)) === null || _a === void 0 ? void 0 : _a.sequenceRecordData().byteLength) || 0; }, instant_world_tracker_anchor_pose_camera_relative: (o, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.instant_world_tracker_anchor_pose_raw(o));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            return res;
        }, instant_world_tracker_anchor_pose: (o, cameraPose, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.instant_world_tracker_anchor_pose_raw(o));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            mat4.multiply(res, cameraPose, res);
            return res;
        }, instant_world_tracker_anchor_pose_set_from_camera_offset: (o, x, y, z, orientation) => {
            // TODO - add an _ext function if we need to support mirrored cameras with an extra param
            // if(mirror) {
            //     x *= -1;
            // }
            // TODO - can we access the appropriate pipeline here to call pipeline_camera_frame_user_facing(p)?
            let userFacing = false;
            let rotation = cameraRotationForScreenOrientation(userFacing) * Math.PI / 180.0;
            let rotationMat = mat4.create();
            mat4.fromRotation(rotationMat, -rotation, [0, 0, 1]);
            let rawVec = vec3.create();
            vec3.transformMat4(rawVec, [x, y, z], rotationMat);
            c.impl.instant_world_tracker_anchor_pose_set_from_camera_offset_raw(o, rawVec[0], rawVec[1], rawVec[2], orientation);
        }, image_tracker_create: pipeline => ImageTracker.create(pipeline, c.impl), image_tracker_destroy: t => { var _a; return (_a = ImageTracker.get(t)) === null || _a === void 0 ? void 0 : _a.destroy(); }, image_tracker_target_type: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_type on a destroyed zappar_image_tracker_t");
                return image_target_type_t.IMAGE_TRACKER_TYPE_PLANAR;
            }
            return obj.getTargetInfo(i).type;
        }, image_tracker_target_count: t => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_count on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.targetCount();
        }, image_tracker_target_load_from_memory: (t, data) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_load_from_memory on a destroyed zappar_image_tracker_t");
                return 0;
            }
            obj.loadFromMemory(data);
        }, image_tracker_target_preview_compressed: (t, i) => {
            var _a;
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_compressed on a destroyed zappar_image_tracker_t");
                return new Uint8Array(0);
            }
            return ((_a = obj.getTargetInfo(i).preview) === null || _a === void 0 ? void 0 : _a.compressed) || new Uint8Array(0);
        }, image_tracker_target_preview_compressed_mimetype: (t, i) => {
            var _a;
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_compressed_mimetype on a destroyed zappar_image_tracker_t");
                return "";
            }
            return ((_a = obj.getTargetInfo(i).preview) === null || _a === void 0 ? void 0 : _a.mimeType) || "";
        }, image_tracker_target_preview_compressed_size: (t, i) => {
            var _a, _b;
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_compressed_size on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return ((_b = (_a = obj.getTargetInfo(i).preview) === null || _a === void 0 ? void 0 : _a.compressed) === null || _b === void 0 ? void 0 : _b.byteLength) || 0;
        }, image_tracker_target_physical_scale_factor: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_physical_scale_factor on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getTargetInfo(i).physicalScaleFactor;
        }, image_tracker_target_radius_top: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_radius_top on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getTargetInfo(i).topRadius;
        }, image_tracker_target_radius_bottom: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_radius_bottom on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getTargetInfo(i).bottomRadius;
        }, image_tracker_target_side_length: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_side_length on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getTargetInfo(i).sideLength;
        }, image_tracker_target_image: (t, i) => {
            let obj = ImageTracker.get(t);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_image on a destroyed zappar_image_tracker_t");
                return new Image();
            }
            return obj.getDecodedPreview(i);
        }, image_tracker_anchor_pose_camera_relative: (o, indx, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.image_tracker_anchor_pose_raw(o, indx));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            return res;
        }, image_tracker_anchor_pose: (o, indx, cameraPose, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.image_tracker_anchor_pose_raw(o, indx));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            mat4.multiply(res, cameraPose, res);
            return res;
        }, image_tracker_target_preview_mesh_indices: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_indices on a destroyed zappar_image_tracker_t");
                return new Uint16Array();
            }
            return obj.getPreviewMesh(indx).indices;
        }, image_tracker_target_preview_mesh_vertices: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_vertices on a destroyed zappar_image_tracker_t");
                return new Float32Array();
            }
            return obj.getPreviewMesh(indx).vertices;
        }, image_tracker_target_preview_mesh_uvs: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_uvs on a destroyed zappar_image_tracker_t");
                return new Float32Array();
            }
            return obj.getPreviewMesh(indx).uvs;
        }, image_tracker_target_preview_mesh_normals: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_normals on a destroyed zappar_image_tracker_t");
                return new Float32Array();
            }
            return obj.getPreviewMesh(indx).normals;
        }, image_tracker_target_preview_mesh_indices_size: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_indices_size on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getPreviewMesh(indx).indices.length;
        }, image_tracker_target_preview_mesh_vertices_size: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_vertices_size on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getPreviewMesh(indx).vertices.length;
        }, image_tracker_target_preview_mesh_uvs_size: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_uvs_size on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getPreviewMesh(indx).uvs.length;
        }, image_tracker_target_preview_mesh_normals_size: (o, indx) => {
            let obj = ImageTracker.get(o);
            if (!obj) {
                zcwarn("attempting to call image_tracker_target_preview_mesh_normals_size on a destroyed zappar_image_tracker_t");
                return 0;
            }
            return obj.getPreviewMesh(indx).normals.length;
        }, face_tracker_anchor_pose_camera_relative: (o, indx, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.face_tracker_anchor_pose_raw(o, indx));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            return res;
        }, face_tracker_anchor_pose: (o, indx, cameraPose, mirror) => {
            let res = applyScreenCounterRotation(undefined, c.impl.face_tracker_anchor_pose_raw(o, indx));
            if (mirror) {
                let scale = mat4.create();
                mat4.fromScaling(scale, [-1, 1, 1]);
                mat4.multiply(res, scale, res);
                mat4.multiply(res, res, scale);
            }
            mat4.multiply(res, cameraPose, res);
            return res;
        }, face_tracker_model_load_default: (o) => __awaiter(this, void 0, void 0, function* () {
            yield loadDefaultFaceModel(o);
        }), face_mesh_create: () => {
            return createFaceMesh();
        }, face_mesh_destroy: (m) => {
            destroyFaceMesh(m);
        }, face_mesh_indices: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_indices on a destroyed zappar_face_mesh_t");
                return new Uint16Array();
            }
            return obj.getIndices();
        }, face_mesh_indices_size: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_indices_size on a destroyed zappar_face_mesh_t");
                return 0;
            }
            return obj.getIndices().length;
        }, face_mesh_uvs: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_uvs on a destroyed zappar_face_mesh_t");
                return new Float32Array();
            }
            return obj.getUVs();
        }, face_mesh_uvs_size: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_uvs_size on a destroyed zappar_face_mesh_t");
                return 0;
            }
            return obj.getUVs().length;
        }, face_mesh_vertices: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_vertices on a destroyed zappar_face_mesh_t");
                return new Float32Array();
            }
            return obj.getVertices();
        }, face_mesh_vertices_size: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_vertices_size on a destroyed zappar_face_mesh_t");
                return 0;
            }
            return obj.getVertices().length;
        }, face_mesh_normals: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_normals on a destroyed zappar_face_mesh_t");
                return new Float32Array();
            }
            return obj.getNormals();
        }, face_mesh_normals_size: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_normals_size on a destroyed zappar_face_mesh_t");
                return 0;
            }
            return obj.getNormals().length;
        }, face_mesh_load_from_memory: (m, ab, fillMouth, fillEyeL, fillEyeR, fillNeck) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_load_from_memory on a destroyed zappar_face_mesh_t");
                return;
            }
            obj.loadFromMemory(ab, fillMouth, fillEyeL, fillEyeR, fillNeck);
        }, face_mesh_update: (m, identity, expression, mirrored) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_update on a destroyed zappar_face_mesh_t");
                return;
            }
            obj.update(identity, expression, mirrored);
        }, face_mesh_load_default: (m) => __awaiter(this, void 0, void 0, function* () {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_load_default on a destroyed zappar_face_mesh_t");
                return;
            }
            let url = new URL("./face_mesh_face_model.zbin", import.meta.url);
            let req = yield fetch(url.toString());
            obj.loadFromMemory(yield req.arrayBuffer(), false, false, false, false);
        }), face_mesh_load_default_face: (m, fillMouth, fillEyeL, fillEyeR) => __awaiter(this, void 0, void 0, function* () {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_load_default_face on a destroyed zappar_face_mesh_t");
                return;
            }
            let url = new URL("./face_mesh_face_model.zbin", import.meta.url);
            let req = yield fetch(url.toString());
            obj.loadFromMemory(yield req.arrayBuffer(), fillMouth, fillEyeL, fillEyeR, false);
        }), face_mesh_load_default_full_head_simplified: (m, fillMouth, fillEyeL, fillEyeR, fillNeck) => __awaiter(this, void 0, void 0, function* () {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_load_default_full_head_simplified on a destroyed zappar_face_mesh_t");
                return;
            }
            let url = new URL("./face_mesh_full_head_simplified_model.zbin", import.meta.url);
            let req = yield fetch(url.toString());
            obj.loadFromMemory(yield req.arrayBuffer(), fillMouth, fillEyeL, fillEyeR, fillNeck);
        }), face_mesh_loaded_version: (m) => {
            let obj = getFaceMesh(m);
            if (!obj) {
                zcwarn("attempting to call face_mesh_load_default on a destroyed zappar_face_mesh_t");
                return -1;
            }
            return obj.getModelVersion();
        }, face_landmark_create: (n) => {
            return createFaceLandmark(n);
        }, face_landmark_destroy: (m) => {
            destroyFaceLandmark(m);
        }, face_landmark_update: (m, identity, expression, mirrored) => {
            let obj = getFaceLandmark(m);
            if (!obj) {
                zcwarn("attempting to call face_landmark_update on a destroyed zappar_face_landmark_t");
                return;
            }
            obj.update(identity, expression, mirrored);
        }, face_landmark_anchor_pose: (m) => {
            let obj = getFaceLandmark(m);
            if (!obj) {
                zcwarn("attempting to call face_landmark_anchor_pose on a destroyed zappar_face_landmark_t");
                return mat4.create();
            }
            return obj.anchor_pose;
        }, html_element_source_create: (pipeline, elm) => HTMLElementSource.createVideoElementSource(pipeline, elm), html_element_source_start: o => { var _a; return (_a = HTMLElementSource.getVideoElementSource(o)) === null || _a === void 0 ? void 0 : _a.start(); }, html_element_source_pause: o => { var _a; return (_a = HTMLElementSource.getVideoElementSource(o)) === null || _a === void 0 ? void 0 : _a.pause(); }, html_element_source_destroy: o => { var _a; return (_a = HTMLElementSource.getVideoElementSource(o)) === null || _a === void 0 ? void 0 : _a.destroy(); }, sequence_source_create: p => SequenceSource.create(p), sequence_source_load_from_memory: (o, data) => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.loadFromMemory(data); }, sequence_source_pause: o => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.pause(); }, sequence_source_start: o => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.start(); }, sequence_source_max_playback_fps_set: (o, fps) => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.maxPlaybackFpsSet(fps); }, sequence_source_time_set: (o, t) => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.setTime(t); }, sequence_source_destroy: o => { var _a; return (_a = SequenceSource.get(o)) === null || _a === void 0 ? void 0 : _a.destroy(); }, permission_granted_all: permissionGrantedAll, permission_granted_camera: permissionGrantedCamera, permission_granted_motion: permissionGrantedMotion, permission_denied_any: permissionDeniedAny, permission_denied_camera: permissionDeniedCamera, permission_denied_motion: permissionDeniedMotion, permission_request_motion: permissionRequestMotion, permission_request_camera: permissionRequestCamera, permission_request_all: permissionRequestAll, permission_request_ui: permissionRequestUI, permission_request_ui_promise: permissionRequestUI, permission_denied_ui: permissionDeniedUI, browser_incompatible: compatibility.incompatible, browser_incompatible_ui: compatibility.incompatible_ui, log_level_set: l => {
            setLogLevel(l);
            c.impl.log_level_set(l);
        } });
    return client;
}
function loadDefaultFaceModel(o) {
    return __awaiter(this, void 0, void 0, function* () {
        let url = new URL("./face_tracking_model.zbin", import.meta.url);
        let data = yield fetch(url.toString());
        let ab = yield data.arrayBuffer();
        client === null || client === void 0 ? void 0 : client.face_tracker_model_load_from_memory(o, ab);
    });
}
